### ISSUE TITLE
Python support: move to python 2.7 instead of 3.4?

### ISSUE THREAD

**Issue by devernay** [MrKepzie/Natron#846]
The whole VFX and animation industry still uses python 2.7, and there is a large codebase available, including assets managers (e.g. https://github.com/shotgunsoftware). Blender is the only known exception.

We will probably switch to 2.7 for the Natron 2.0 release (which is coming soon). I there are any strong objections, please file them here for discussion.

**Comment by blackearth2014** [comment]
We know Blender is a class and culture by itself. So Blender is not an issue. Natron is in a class and culture by itself as well, but will be use by more diverse production houses that still uses Python 2.7. When version 3 is ready to be release, then it should have Python 3.4 or higher for PyNode developers for Natron. I personally feel that pushing the development envelope with the latest API's will allow new creative and passionate developers to contribute to Natron a lot more. 

Natron is still young but yet very powerful. By version 3, we shall be more mature and starting to set standards for the open source community. Commercial apps need not apply.

Just my half a penny cents.

**Comment by olear** [comment]
If the majority of users/applications still uses Python2 then we should adapt in my opinion.

We should confirm as much as possible to the CY2015 standard (http://www.vfxplatform.com/) (Nuke is at CY2014). And if you see CY2016 (draft?) they are still on Python2.

**Comment by azerupi** [comment]
> The whole VFX and animation industry still uses python 2.7

I would argue that the advantage of Natron is that it doesn't need to be held back by the industry standards. I think industry standards are a good thing, but can sometimes prevent improvement. As an open source project I think Natron could and should be on the "bleeding edge" and push the standards.

Python 3 has been around since 2008 (7 years) I think it's fair enough for Natron to only support Python 3.

Now I am not in the industry myself and I can understand if it is mandatory for integration in current workflows. But it's also worth considering if Natron is, or will be widely used in the industry in the short term future. 

It's just my opinion though as a hobbyist and a developer (who hates to be forced to use python2) :wink:

**Comment by blackearth2014** [comment]
Azerupi is right. I thought Python 3.+ was fairly new. If 7 years it's been out, then go directly to Python 3.4 without concerns of being backwards compatible with old tech. Stay on the blending edge and keep innovating, let everyone else catch up.

**Comment by olear** [comment]
How old something is doesn't matter, compatibility does. Innovation should happen in the app/plugins, not in third-party (or let me rephrase, innovation should be independent of third-party). What does Python3 offer that so great compared to Python2? Or what can't you do in Python2 that you can in Python3? (I don't care, but if the only reason is "old" than that's not good enough in my opinion).

On Linux I build against the lowest (sane) possible libraries to be able to make a version that runs on almost any Linux flavour, if I was to go "bleeding edge" /or don't care about "legacy" then the Linux version would only work on a couple of distros (at best). We should also cut support for Windows 7/8 they are "old", Win10 for everyone  .......... now I'm off-topic... :) 

Also, if/when Natron moves to Python2 everybody are free to rebuild against Python3 if they want ;)

**Comment by MrKepzie** [comment]
Unfortunately it doesn't work that way, many studios have tools built against Python 2.7 and forcing them to use Python 3 is a showstopper, so the only solution for now if want to have industry appreciation is to stick with what the vfxplatform suggests.

**Comment by azerupi** [comment]
If Natron switches to Python 2 does that mean that plug-ins will have to be written in Python 2?
That is actually my main concern.


> many studios have tools built against Python 2.7 and forcing them to use Python 3 is a showstopper

Sure, like a said I am not in the industry. If it is mandatory then Natron should do it. 

But are studios really going to use Natron in the short term future? (it's a legitimate question) 
If they stick to the industry standards an already have tools I don't see them use Natron in the next couple of years.

> if I was to go "bleeding edge" /or don't care about "legacy" then the Linux version would only work on a couple of distros (at best).

**Comment by Ambassy** [comment]
To shift to pyton 3.4 was a good idea,but if almost industry standard tools are on 2.7 the shift to pyton 3 will make pros reject natron out. In my opinion, Stick to industry standard(2.7) and move to pyton 3 when the industry also moves on pyton 3.

**Comment by olear** [comment]
CY2017 is open for suggestions/feedback, but that's a long time from now (v4ish).

**Comment by devernay** [comment]
Also, from http://www.vfxplatform.com/FAQ/ :
"*When will the Reference Platform move to Python 3?*
This is a complex issue given the lack of backwards compatibility to Python 2.6/2.7. The right time to make the switch will decided by the software vendors in consultation with their customers and it’s possible Python 3 may not see wide adoption in VFX until and unless Python 3 is fully compatible with all the legacy Python code in active use across the industry."

To support both versions in your scripts, just use print with parentheses, and when you divide two ints and expect the result to be a float , cast the denominator to float. If you actually use integer division in your code, it may be harder.
See http://python3porting.com/noconv.html for example.

AFAIK there is no Python3 library/module which doesn't exist in Python2.

The problem may also be reformulated as "should Python3 have ever existed?".

I'm using Python3 in my classes, and frankly this is a pain in the ass.

**Comment by blackearth2014** [comment]
Do you teach Python Fred in a school somewhere.

**Comment by rcspam** [comment]
To the extent that Natron will replace at first some commercial softwares use [in some graphical high-schools](http://libregraphicsworld.org/blog/entry/the-complete-story-of-paris-8-university-going-for-krita-blender-natron). Industry standard must be respect and Natron grow up with it. So python 2.7 should be use. There are not a lot  python Natron plugins at this moment, so although i'm not an python expert, it should not be too difficult for the writers to downgrade what little there is.

**Comment by mestaritonttu** [comment]
This is a classic chicken or the egg -dilemma.

The reference to Shotgun Software as an example of the problem turns out to be circular logic: in their python-api repo they have an issue dated 2013 about Python 3 where they respond:
> Python 3 support is likely to happen eventually but to be honest the demand has been pretty non-existant since most studios and software still relies on 2.x.

Nothing is stopping studios from using Python 2 & 3 in the same environment (easy with tools like https://github.com/yyuu/pyenv) and they certainly realize this transition must take place sooner or later (Python 2.7 EOL is 2020).

The question was always "why should we switch to Python 3, when it breaks bc and doesn't give us any hot new features?" As v3 matures, though, this line of reasoning [is becoming obsolete](http://nafiulis.me/python-3-may-become-relevant-now.html).

I don't really understand VFXplatform's "Python 3 should become compatible with legacy code" -expectation! Maybe it is referring to dependencies on libraries that haven't yet released a Python 3 -compatible version. [The situation with the top 200 libs is looking pretty good, though](http://python3wos.appspot.com/).

If I were you, I would gather more data. Will studios really want to plug their legacy code somehow into Natron? What are the plans of transitioning to Python 3 in studios & relevant software products? These types of stats.

**Comment by mash-graz** [comment]
python3 has a much cleaner unicode handling than earlier versions.
that's the main difference from my point of view.

if you are just doing image processing and adding a few lines of foreign language comments to the code, this does not matter so much. but if you have to handle clean encoding in databases and nowadays communication protocols it changes a lot.
i wouldn't go back anymore... 
and porting 2.x code to 3.x is so easy...

**Comment by nicholascarroll** [comment]
Don't DOWNGRADE!
Just wait and industry will catch up.

**Comment by olear** [comment]
That's sometime in the future... (at least two years)

And you can build Natron yourself with whatever Python version wanted.

**Comment by mash-graz** [comment]
just a remark from an insider:

http://www.snarky.ca/why-python-3-exists

**Comment by olear** [comment]
As said in August, you are free to build Natron with Python 3 support if you want...

We will follow the guidelines from http://www.vfxplatform.com/, you can express your opinions regarding the specs at https://groups.google.com/forum/m/#!forum/vfx-platform-discuss

**Comment by mash-graz** [comment]
> As said in August, you are free to build Natron with Python 3 support if you want...

compiling the with an arbitrary version of python isn't hard, but debugging all the related nasty runtime errors looks much more complicated.

sure -- there exists some compatibility solutions (six/python-future etc.) as a work around, but in fact it's a very important design decision that could minimize foreseeable troubles in the future. simply ignoring this kind of serious issues is the worst way to go.

> We will follow the guidelines from http://www.vfxplatform.com/, ...

usually i really like open standards and attempts to advance compatibility, but this recommendations in particular are one of the worst examples how to do it. :(
it often seems to me, they just got formulated, to build an excuse for very sluggish and conservative commercial products, to ignore all user requests (e.g. for more recent OpenEXR libary and compression options) and neglect the real needs for compatibility in the context of a much faster growing free software ecosystem and it's enormous speed of development. IMHO recommendations and warnings from upstream developers of all the affected subsystems should be taken at least as important! otherwise this vfxplatform recommendations are the perfect recipe to become less innovative and user unfriendly as all this well known huge commercial alternatives.
